substratest/assets.py

```diff
@@ -168,15 +139,54 @@ def load(cls, d):
             raise errors.TError(f"cannot parse asset `{d}`") from e
 
 
-class _Asset(_DataclassLoader, abc.ABC):
+class _InternalStruct(pydantic.BaseModel, _DataclassLoader, abc.ABC):
+    """Internal nested structure"""
+
+
+class _Asset(_InternalStruct, abc.ABC):
     """Represents assets stored in the Substra framework.
 
     Convert a dict with camel case fields to a Dataclass.
     """
 
 
-@dataclasses.dataclass(frozen=True)
-class Permission(_DataclassLoader):
+class _BaseFutureAsset(_Asset):
+    __session: typing.Any = None
+    _future_cls = None
+
+    def attach(self, session):
+        """Attach session to asset."""
+        object.__setattr__(self, '__session', session)
+        return self
+
+    @property
+    def _session(self):
+        # because self.__session doesn't work properly with Pydantic, we have to use
+        # __getattribute__ and __setattr__ (https://docs.python.org/3/reference/datamodel.html)
```

Comment:
maybe saying that "becausse Pydantic doesn't support private fields" is clearer than saying "self.__session doesn't work" ?

Reply:
Yes, I will change this as well, it isn't really clear indeed